### PR TITLE
Accept Annex B octal up to \377 and \8, \9 literals in RegExp character classes

### LINKED ISSUE #1275: Annex B RegExp character class decimal escape in TypeScript 2.1.4
Normally I don't report real-world regexp breakages anymore because I know that nonstandard regexp support is still a work in progress.  However, since this breaks the TypeScript compiler it seemed worthwhile to call it out.  The offending regexp is:
```js
var tripleSlashDirectiveFragmentRegex = /^(\/\/\/\s*<reference\s+(path|types)\s*=\s*(?:'|"))([^\3"]*)$/;
```

I don't know if there are more broken regexps, but that's the one that caused an error for me.  It was a bit too convoluted for me to try to find the offending token :)

### PR DESCRIPTION
**Pull request by svaarala**
Tasks:
- [x] Initial hack to test parsing approach
- [x] Refactor so that legacy octal parsing is shared between normal lexer and regexp charclass lexer
- [x] Testcases (from #1275)
- [x] Releases entry
- [x] Duktape-wiki post-es5 feature => RegExp Annex B syntax features are not described individually

Fixes #1275.

### CODE REVIEW

**Comment by svaarala** [comment]
Just a quick hack test which seems to work, I'll rework and add tests later today.

File: tests/ecmascript/test-lex-legacy-octal-regexp.js [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+ *  Legacy octal literals like \377 in regexps.
+ */
+
+var re;
+
+/*===
+true false
+===*/
+
+// \0 is interpreted as a literal zero, even without legacy support.
```
Comment (fatcerberus) [inline tests/ecmascript/test-lex-legacy-octal-regexp.js]: I assume you mean "interpreted as NUL" here?

Comment (svaarala) [inline tests/ecmascript/test-lex-legacy-octal-regexp.js]: Yeah, I'll fix.

**Comment by svaarala** [comment]
Seems to be accepted by Node.js in strict mode.

**Comment by svaarala** [comment]
To be clear, the code diff is just a hack (comments/debug prints are copy-paste etc) for now.

**Comment by fatcerberus** [comment]
> To be clear, the code diff is just a hack (comments/debug prints are copy-paste etc) for now.

I noticed, don't worry. :)

In any case it does seem to be working as intended.  The TypeScript compiler runs without manual modification, at least.

**Comment by svaarala** [comment]
Good to hear :)

**Comment by fatcerberus** [comment]
Interestingly enough, Node.js:
```
> (function() { 'use strict'; return /\3/; })()
/\3/
```

Seems to accept octal escapes in regexps in strict mode.  Strings, however, are rejected:
```
> (function() { 'use strict'; return "\3"; })()
(function() { 'use strict'; return "\3"; })()

SyntaxError: Octal literals are not allowed in strict mode.
```

So not sure whether V8 is out of spec or if that's the correct behavior.

**Comment by fatcerberus** [comment]
That's with Node 7.8.0, by the way.

**Comment by svaarala** [comment]
As far as I can tell the octal literals etc in strings are explicitly forbidden in strict mode, Annex B text:

> The syntax and semantics of 11.8.3 is extended as follows except that this extension is not allowed for strict mode code: [...]

But no strict mode requirements exists for additional RegExp syntax. So I'd say it's a feature.

**Comment by svaarala** [comment]
@fatcerberus I'm done with the code changes and I'll merge this shortly unless something comes up :)

**Comment by fatcerberus** [comment]
That was quick :)

**Comment by fatcerberus** [comment]
I guess the reason legacy octal is allowed in regexps even in strict mode is for compatibility with other (non-JS) regex engines.  It would be annoying to copy a regex from elsewhere and maybe have it rejected just because the code happens to be in strict mode.  That's less of an issue for, e.g., string literals, since their syntax has no implicit meaning outside the context of Ecmascript.

**Pull request merged**